src/os/bluestore/BlueFS.cc

```diff
@@ -3495,7 +3789,28 @@ void BlueFS::flush_range(FileWriter *h, uint64_t offset, uint64_t length)/*_WF*/
 {
   _maybe_check_vselector_LNF();
   std::unique_lock hl(h->lock);
-  _flush_range_F(h, offset, length);
+  if (h->file->is_new_wal()) {
+    //For wal files disregard offset and length and just flush current envelope.
+    _flush_wal_F(h);
+  } else {
+    _flush_range_F(h, offset, length);
+  }
+}
+
+int BlueFS::_flush_wal_F(FileWriter *h)
+{
+  ceph_assert(ceph_mutex_is_locked(h->lock));
+  ceph_assert(h->file->is_new_wal());
+  uint64_t content_length = h->get_buffer_length() - sizeof(File::wal_flush_t::wal_length_t);
+  h->append((char*)&h->file->wal_marker.v[0], 8);
+  uint64_t offset = h->pos;
+
+  h->file->fnode.wal_size += content_length;
+  ceph_le64 flush_length_le(content_length);
+  h->wal_header_filler.copy_in(sizeof(flush_length_le), (char*)&flush_length_le);
+  uint64_t length = sizeof(File::wal_flush_t::wal_length_t) + content_length + sizeof(h->file->wal_marker);
```

Comment:
Wouldn't it be better to use head_size()/tail_size() helpers instead of sizeof?

Reply:
Replaced all sizeofs with header_size()  / tail_size() calls.